documentation/puppet_server_metrics_performance.markdown

```diff
@@ -0,0 +1,78 @@
+
+Several factors can limit your Server infrastructure's ability to produce JRubies.
+
+### Request-handling capacity
+
+If your free JRubies are 0 or fewer, your server is receiving more requests for JRubies than it can provide, which means it must queue those requests to wait until resources are available. Puppet Server performs best when the average number of free JRubies is above 1, which means Server always has enough resources to immediately handle incoming requests.
+
+There are two indicators in Puppet Server's metrics that can help you identify a request-handling capacity issue:
+
+-   **Average JRuby Wait Time:** This refers to the amount of time Puppet Server has to wait for an available JRuby to become available, and increases when each JRuby is held for a longer period of time, which reduces the overall number of free JRubies and forces new requests to wait longer for available resources.
+-   **Average JRuby Borrow Time:** This refers to the amount of time that Puppet Server "holds" a JRuby as a resource for a request, and increases because of other factors on the server.
+
+If wait time increases but borrow time stays the same, your Server infrastructure might be serving too many agents. This indicates that Server can easily handle requests but is receiving too many at once to keep up.
+
+If both wait and borrow times are increasing, something else on your server is causing requests to take longer to process. The longer borrow times suggest that Puppet Server is struggling more than before to process requests, which has a cascading effect on wait times. Correlate borrow time increases with other events whenever possible to isolate what activities might cause them, such as a Puppet code change.
+
+If you are setting up Puppet Server for the first time, start by increasing your Server infrastructure's capacity through additional JRubies (if your server has spare CPU and memory resources) or compile masters until your free JRubies are at least 0 and average number of free JRubies are at least 1. Once your system can handle its request volume, you can start looking into more specific performance improvements.
+
+#### Adding more JRubies
+
+If you need to add JRubies, remember that Puppet Server is tuned by default to use one fewer than your total number of CPUs, with a maximum of 4 CPUs, for the number of available JRubies. You can change this by setting `max-active-instances` in [`puppetserver.conf`][], under the `jruby-puppet` section. Don' set this value to something greater than the number of CPUs, because if there aren't enough processes available, it won't matter how many JRubies in the pool there are — the server won't be able to use them.
```

Comment:
Maybe it would be worth suggesting that since catalog requests, the most expensive ones that agents make, are CPU-bound that it's typically best to start with a `max-active-instances` value which is no greater than the number of CPU cores on the host machine.  For cases where the catalogs that the master is compiling lead to a large number of follow-up "file" requests (file_content, file_metadata, etc.) which are more I/O than CPU bound, it may be possible to improve performance further by setting `max-active-instances` to a value greater than the number of available CPU cores.

@npwalker, since I think you have more relevant real-world experience with this than we do, any specific recommendations around the wording that you'd suggest here?